src/Tools/ExternalAccess/RazorTest/Cohost/RazorCohostTests.cs

```diff
@@ -0,0 +1,233 @@
+        {
+            TextDocument = new VersionedTextDocumentIdentifier
+            {
+                Uri = document.GetURI()
+            },
+            ContentChanges =
+            [
+                new TextDocumentContentChangeEvent
+                {
+                    Range = new VisualStudio.LanguageServer.Protocol.Range
+                    {
+                        Start = new Position(0, 0),
+                        End = new Position(0, 0)
+                    },
+                    Text = "Not The "
+                }
+            ]
+        };
+
+        await server.GetTestAccessor().ExecuteRequestAsync<DidChangeTextDocumentParams, object>(Methods.TextDocumentDidChangeName, didChangeRequest);
+
+        Assert.True(workspaceManager.GetTrackedLspText().TryGetValue(document.GetURI(), out trackedText));
+        Assert.Equal("Not The Original text", trackedText.Text.ToString());
+    }
+
+    private TestWorkspace CreateWorkspace(string workspaceXml)
+    {
+        var testWorkspace = CreateWorkspace(options: null, mutatingLspWorkspace: false, workspaceKind: null);
+        testWorkspace.InitializeDocuments(XElement.Parse(workspaceXml), openDocuments: false);
+        return testWorkspace;
+    }
+
+    private static async Task<AbstractLanguageServer<RequestContext>> InitializeLanguageServerAsync(TestWorkspace testWorkspace)
+    {
+        var languageClient = testWorkspace.ExportProvider.GetExportedValues<ILanguageClient>().OfType<RazorCohostLanguageClient>().Single();
+        await languageClient.ActivateAsync(CancellationToken.None);
+
+        var server = languageClient.GetTestAccessor().LanguageServer;
+        Assert.NotNull(server);
+
+        var serverAccessor = server!.GetTestAccessor();
+
+        await serverAccessor.ExecuteRequestAsync<InitializeParams, InitializeResult>(Methods.InitializeName, new InitializeParams { Capabilities = new() });
+
+        return server;
+    }
+
+    internal class TestRequest(Uri documentUri, Guid projectId)
+    {
+        public Uri DocumentUri => documentUri;
+        public Guid ProjectId => projectId;
+    }
+
+    [LanguageServerEndpoint(MethodName)]
+    [ExportRazorStatelessLspService(typeof(RazorHandler)), Shared]
+    [method: ImportingConstructor]
+    [method: Obsolete("This exported object must be obtained through the MEF export provider.", error: true)]
+    internal class RazorHandler() : AbstractRazorCohostDocumentRequestHandler<TextDocumentPositionParams, TestRequest>
+    {
+        internal const string MethodName = "testMethod";
+
+        protected override bool MutatesSolutionState => false;
+
+        protected override bool RequiresLSPSolution => true;
+
+        protected override RazorTextDocumentIdentifier? GetRazorTextDocumentIdentifier(TextDocumentPositionParams request)
+        {
+            return new RazorTextDocumentIdentifier(request.TextDocument.Uri, (request.TextDocument as VSTextDocumentIdentifier)?.ProjectContext?.Id);
```

Comment:
All of the endpoints in the Razor repo will have to do this, in particular the second parameter, in order to ensure project context flows correctly. Not too bad, and I couldn't see a way around it.